### PR TITLE
upsert action now uses the engine directly instead of using the index transport

### PR DESCRIPTION
**Pull request by seut**
This is a huge performance improvement when replicas are configured, because now 1 replica request is made instead 1 for each bulk item.
related benchmark results: https://gist.github.com/seut/183a8a69c6b6491ac873

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/TransportShardAction.java [hunk 1]
```diff
@@ -107,7 +106,8 @@ public Tuple call() throws Exception {
             //noinspection unchecked
             response = callable.call();
         } catch (Throwable e) {
-            throw Throwables.propagate(e);
+            assert e instanceof RuntimeException : "Expecting instance of RuntimeException here";
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/TransportShardAction.java]: this looks odd, was this just for debugging or why is this important here?

File: sql/src/main/java/io/crate/jobs/AbstractExecutionSubContext.java [hunk 2]
```diff
@@ -41,6 +41,7 @@
 
     protected AbstractExecutionSubContext(int id) {
         this.id = id;
+        logger = Loggers.getLogger(this.getClass());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/AbstractExecutionSubContext.java]: Would it make sense to add a logger as argument to the constructor instead? `getLogger` can be fairly expensive

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java [hunk 3]
```diff
@@ -202,6 +209,7 @@ private ShardId shardId(String indexName, String id, @Nullable String routing) {
             ).shardId();
         } catch (IndexMissingException e) {
             if (!autoCreateIndices) {
+                indicesDeleted.add(indexName);
```
Comment (mfussenegger) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java]: the exception is thrown in the next line which probably causes the whole process to be aborted anyway, why is it important to add the indexName to indicesDeleted here?

File: sql/src/main/java/io/crate/executor/transport/TransportShardAction.java (hunk 1 above)
Comment (seut) [inline sql/src/main/java/io/crate/executor/transport/TransportShardAction.java]: wrapping any throwable into a runtime exception does not work here as all method usages does not honor wrapped exceptions (e.g. the retry logic)

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java (hunk 3 above)
Comment (seut) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java]: index re-creation must be prevented while adding new items. so this set will be checked on add(..)

File: sql/src/main/java/io/crate/executor/transport/TransportShardAction.java (hunk 1 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/TransportShardAction.java]: I don't get it =)
Throwables.propagate(e) doesn't wrap the exception if it already is a RuntimeException

File: sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java (hunk 3 above)
Comment (mfussenegger) [inline sql/src/main/java/org/elasticsearch/action/bulk/BulkShardProcessor.java]: Isn't there anyway a window for the race condition? Imho this is only "optimizing" the delete race condition to be less expensive. Not sure if this is worth it.

File: sql/src/test/java/io/crate/integrationtests/UpdateIntegrationTest.java [hunk 4]
```diff
@@ -608,6 +630,7 @@ public void testUpdateRetryOnVersionConflict() throws Exception {
         execute("insert into test (a, b) values ('foo', 1)");
         assertThat(response.rowCount(), is(1L));
         refresh();
+        execute("select _version from test");
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/UpdateIntegrationTest.java]: result of this query isn't used

Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/UpdateIntegrationTest.java]: ah yeah, debugging relicts, thx

**Comment by mfussenegger** [comment]
ack

**Pull request merged**